### ISSUE TITLE
Where does the empty transaction come from?

### ISSUE THREAD

**Issue by peter-gribanov** [doctrine/doctrine2#7175]
![image](https://user-images.githubusercontent.com/1954436/38430361-b6fca4a6-39c9-11e8-9fa7-7a3824511a58.png)

I see that two requests are executed on the page, but they do not make sense.
I don't create a transaction explicitly.

Perhaps this is due to this code in my project:

```php
use Doctrine\Common\Persistence\ManagerRegistry;
use Doctrine\ORM\EntityManagerInterface;
use Symfony\Component\Console\ConsoleEvents;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;
use Symfony\Component\HttpKernel\Event\PostResponseEvent;
use Symfony\Component\HttpKernel\KernelEvents;

class FlusherSubscriber implements EventSubscriberInterface
{
    private $registry;

    public function __construct(ManagerRegistry $registry)
    {
        $this->registry = $registry;
    }

    public static function getSubscribedEvents()
    {
        return [
            ConsoleEvents::TERMINATE => 'onTerminate',
            KernelEvents::TERMINATE => 'postResponse',
        ];
    }

    public function onTerminate()
    {
        $this->flush();
    }

    public function postResponse(PostResponseEvent $event)
    {
        if ($event->isMasterRequest()) {
            $this->flush();
        }
    }

    private function flush()
    {
        foreach ($this->registry->getManagers() as $manager) {
            /* @var $manager EntityManagerInterface */
            if ($manager->getConnection()->isConnected() && $manager->isOpen()) {
                $manager->flush();
            }
        }
    }
}
```

**Comment by Ocramius** [comment]
This would need an integration test to be figured out, sorry...

**Comment by zquintana** [comment]
Look at `postResponse` that will execute a flush without first checking that it's open will execute those empty transactions. Kind of a neat idea your doing here, never seen someone try leverage application events to control entity manager state.

**Comment by Ocramius** [comment]
A `flush()` on an empty `UnitOfWork` shouldn't start a transaction (not even a DB connection):


https://github.com/doctrine/doctrine2/blob/v2.6.1/lib/Doctrine/ORM/UnitOfWork.php#L359

**Comment by peter-gribanov** [comment]
> Look at postResponse that will execute a flush without first checking that it's open will execute those empty transactions.

@zquintana I not only check that the connection is open, but i also check that it is not closed. Otherwise, you can get an incomprehensible error at [this](https://github.com/doctrine/doctrine2/blob/99db207a9f1214ad483efd81341279a550526358/lib/Doctrine/ORM/EntityManager.php#L356) line.

> Kind of a neat idea your doing here, never seen someone try leverage application events to control entity manager state.

@zquintana The advantage of the Doctrine is that we can perform the flush only once per request and not call it after each persist or change in entities.

**Comment by zquintana** [comment]
@Ocramius forgot that, your right. @peter-gribanov, yes you can do it that way.  What version of Doctrine are you running? Have you customized it in anyway? Also do you have listeners setup on the entity manager?

**Comment by Ocramius** [comment]
> @zquintana I not only check that the connection is open, but i also check that it is not closed. Otherwise, you can get an incomprehensible error at [this](https://github.com/doctrine/doctrine2/blob/99db207a9f1214ad483efd81341279a550526358/lib/Doctrine/ORM/EntityManager.php#L356) line.

What's incomprehensible about that? Can you maybe suggest an improvement?

**Comment by peter-gribanov** [comment]
> What's incomprehensible about that? Can you maybe suggest an improvement?

@Ocramius it's not a question. I just commented.

**Comment by Ocramius** [comment]
> I just commented.

Yes, which is why I asked for a possible improvement there. Keeping things as they are and "commenting" won't change anything for the better.

**Comment by Ocramius** [comment]
This sort of info is still insufficient to figure out what's going on.

You need to create an isolated test case that reproduces your scenario first. A tip would be to use the existing service configuration from your current DoctrineBundle setup, and then remove pieces one by one until the test doesn't fail anymore.

That will pin-point the issue and allow the test to be set up with a failing config as well.

**Comment by peter-gribanov** [comment]
Fuck. A bug exists, but it is very difficult to reproduce it.
On the same page, this error may or may not be.

And unfortunately, i can not reproduce the problem on an empty project.

The problem occurs on pages when there are no changes.
All queries are only for select data.

For analysis, i added a debug code [here](https://github.com/doctrine/doctrine2/blob/v2.6.1/lib/Doctrine/ORM/UnitOfWork.php#L705).

```php
// if regular field
if ( ! isset($class->associationMappings[$propName])) {
    if ($isChangeTrackingNotify) {
        continue;
    }

    var_dump(get_class($orgValue));
    var_dump($orgValue == $actualValue);
    var_dump($orgValue === $actualValue);

    var_dump(spl_object_id($orgValue));
    var_dump(spl_object_id($actualValue));

    $ref = new \ReflectionClass(get_class($orgValue));
    foreach ($ref->getProperties() as $property) {
        $property->setAccessible(true);
        var_dump($property->getValue($orgValue) === $property->getValue($actualValue));
    }

    $changeSet[$propName] = [$orgValue, $actualValue];

    continue;
}
```

and i see this:

```
string(49) "App\Contest\Domain\Model\Profile"
bool(true)
bool(false)
int(7290)
int(7312)
bool(true)
bool(true)
bool(true)
bool(true)
bool(true)
string(47) "App\Contest\Domain\Model\Decor"
bool(true)
bool(false)
int(7291)
int(7318)
bool(true)
bool(true)
string(53) "App\Contest\Domain\Model\JoinOptions"
bool(true)
bool(false)
int(7292)
int(7324)
bool(true)
bool(true)
bool(true)
bool(true)
string(53) "App\Contest\Domain\Model\VoteOptions"
bool(true)
bool(false)
int(7295)
int(7330)
bool(true)
bool(true)
bool(true)
```

That is, they are different instances of the object, but they are equal.

All changes are Embedded and each Embedded in the class is considered to be changed.

I repeat. I don't change this object, but UnitOfWork still believes that it is changed.

Any idea where else to dig?

**Comment by Ocramius** [comment]
As I said above:

 > Can you make a reproducible scenario as I asked above?

**Comment by Ocramius** [comment]
As I said above:

 > Can you make a reproducible scenario as I asked above?

**Comment by Ocramius** [comment]
As I said above:

 > Can you make a reproducible scenario as I asked above?

**Comment by peter-gribanov** [comment]
As I said above:

> i can not reproduce the problem on an empty project.

On the empty project does not reproduce.
On the my project on the local mashine it is very rare reproduce.
On the my project on the production server is reproduce often.

For obvious reasons, i can not provide a my project for analysis.
I will try to reproduce the problem on the [empty project](https://github.com/peter-gribanov/doctrine_empty_transaction).

**Comment by Ocramius** [comment]
@peter-gribanov we need something in https://github.com/doctrine/doctrine2/tree/46b695d67cc77e5d4ad49a864dbdbc249b965ef4/tests/Doctrine/Tests/ORM/Functional/Ticket, not in an example project.

**Comment by Michal-sk** [comment]
Not sure if related, but im also having blank transactions with Embeddables, the embeddable is always marked as changed.

@peter-gribanov  are you using marc-mabe/php-enum ?

My issue: https://github.com/doctrine/doctrine2/issues/7204

**Comment by peter-gribanov** [comment]
@Michal-sk [marc-mabe/php-enum](https://github.com/marc-mabe/php-enum), [myclabs/php-enum](https://github.com/myclabs/php-enum) and [happy-types/enumerable-type](https://github.com/antanas-arvasevicius/enumerable-type) has several disadvantages and so i use [my implementation](https://github.com/gpslab/enum) for enum. 
But this is not related to this problem because not all my Embeddables contain custom types.

**Comment by peter-gribanov** [comment]
> Not sure if related, but im also having blank transactions with Embeddables, the embeddable is sometimes marked as changed even when no actual / obvious change happend

Quite right. It is because of this that an empty transaction appears.

**Comment by Michal-sk** [comment]
Sure, but why i'm asking is not because of custom types but more that the instances are not the same anymore when unserialize happens, and that's why the identical check would fail.

Please keep this issue updated if you find something. Might be related to my issue as well.
